From 21e2307a16d13faa5a866bb72735bf94e142c305 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Wed, 2 Apr 2014 16:47:51 +0900
Subject: [PATCH 10/16] vspfilter: Wrap the frame transformation function to
 pass a GstVspFilterFrame union

The GstVspFilterFrame union shares a GstVideoFrame and a dmabuf descriptor.

This is preliminary for the dmabuf handling support.
---
 gst/vspfilter/gstvspfilter.c | 28 +++++++++++++++++++++-------
 gst/vspfilter/gstvspfilter.h |  6 ++++++
 2 files changed, 27 insertions(+), 7 deletions(-)

diff --git a/gst/vspfilter/gstvspfilter.c b/gst/vspfilter/gstvspfilter.c
index e8d0cd6..945b7ab 100644
--- a/gst/vspfilter/gstvspfilter.c
+++ b/gst/vspfilter/gstvspfilter.c
@@ -93,6 +93,8 @@ static gboolean gst_vsp_filter_set_info (GstVideoFilter * filter,
     GstVideoInfo * out_info);
 static GstFlowReturn gst_vsp_filter_transform_frame (GstVideoFilter * filter,
     GstVideoFrame * in_frame, GstVideoFrame * out_frame);
+static GstFlowReturn gst_vsp_filter_transform_frame_process (GstVideoFilter *
+    filter, GstVspFilterFrame in_vframe, GstVspFilterFrame out_vframe);
 
 /* copies the given caps */
 static GstCaps *
@@ -1307,8 +1309,8 @@ start_capturing (GstVspFilter * space, int fd, int index,
 }
 
 static GstFlowReturn
-gst_vsp_filter_transform_frame (GstVideoFilter * filter,
-    GstVideoFrame * in_frame, GstVideoFrame * out_frame)
+gst_vsp_filter_transform_frame_process (GstVideoFilter * filter,
+    GstVspFilterFrame in_vframe, GstVspFilterFrame out_vframe)
 {
   GstVspFilter *space;
   GstVspFilterVspInfo *vsp_info;
@@ -1339,7 +1341,7 @@ gst_vsp_filter_transform_frame (GstVideoFilter * filter,
    */
   if (!set_vsp_entities (space, in_info->finfo->format, in_info->width,
           in_info->height, out_info->finfo->format,
-          GST_VIDEO_FRAME_COMP_STRIDE (out_frame, 0) /
+          GST_VIDEO_FRAME_COMP_STRIDE (out_vframe.frame, 0) /
           out_info->finfo->pixel_stride[0], out_info->height)) {
     GST_ERROR_OBJECT (space, "set_vsp_entities failed");
     return GST_FLOW_ERROR;
@@ -1347,14 +1349,14 @@ gst_vsp_filter_transform_frame (GstVideoFilter * filter,
 
   /* set up planes for queuing an input buffer */
   for (i = 0; i < vsp_info->n_planes[OUT]; i++) {
-    in_planes[i].m.userptr = (unsigned long) in_frame->map[i].data;
-    in_planes[i].length = in_frame->map[i].size;
+    in_planes[i].m.userptr = (unsigned long) in_vframe.frame->map[i].data;
+    in_planes[i].length = in_vframe.frame->map[i].size;
   }
 
   /* set up planes for queuing an output buffer */
   for (i = 0; i < vsp_info->n_planes[CAP]; i++) {
-    out_planes[i].m.userptr = (unsigned long) out_frame->map[i].data;
-    out_planes[i].length = out_frame->map[i].size;
+    out_planes[i].m.userptr = (unsigned long) out_vframe.frame->map[i].data;
+    out_planes[i].length = out_vframe.frame->map[i].size;
   }
 
   queue_buffer (space, vsp_info->v4lout_fd, OUT,
@@ -1403,6 +1405,18 @@ gst_vsp_filter_transform_frame (GstVideoFilter * filter,
   return GST_FLOW_OK;
 }
 
+static GstFlowReturn
+gst_vsp_filter_transform_frame (GstVideoFilter * filter,
+    GstVideoFrame * in_frame, GstVideoFrame * out_frame)
+{
+  GstVspFilterFrame in_vframe, out_vframe;
+
+  in_vframe.frame = in_frame;
+  out_vframe.frame = out_frame;
+
+  return gst_vsp_filter_transform_frame_process (filter, in_vframe, out_vframe);
+}
+
 static gboolean
 plugin_init (GstPlugin * plugin)
 {
diff --git a/gst/vspfilter/gstvspfilter.h b/gst/vspfilter/gstvspfilter.h
index 722bd93..3fa196d 100644
--- a/gst/vspfilter/gstvspfilter.h
+++ b/gst/vspfilter/gstvspfilter.h
@@ -65,6 +65,7 @@ typedef struct _GstVspFilter GstVspFilter;
 typedef struct _GstVspFilterClass GstVspFilterClass;
 
 typedef struct _GstVspFilterVspInfo GstVspFilterVspInfo;
+typedef union _GstVspFilterFrame GstVspFilterFrame;
 
 struct buffer {
   void *start;
@@ -99,6 +100,11 @@ struct _GstVspFilterVspInfo {
   guint16 plane_stride[MAX_DEVICES][VIDEO_MAX_PLANES];
 };
 
+union _GstVspFilterFrame {
+  GstVideoFrame *frame;
+  gint dmafd;
+};
+
 /**
  * GstVspFilter:
  *
-- 
1.8.0

